src/approaches/grammar_search_invention_approach.py

```diff
@@ -702,6 +705,82 @@ def _evaluate_atom_trajectory(self, atoms_sequence: List[Set[GroundAtom]],
             score += -trans_log_prob  # remember that lower is better
         return score
 
+
+@dataclass(frozen=True, eq=False, repr=False)
+class _HAddHeuristicBasedScoreFunction(_HeuristicBasedScoreFunction):  # pylint:disable=abstract-method
+    """Implement _generate_heuristic() with HAdd.
+    """
+    def _generate_heuristic(self, init_atoms: Set[GroundAtom],
+                            objects: Set[Object],
+                            goal: Set[GroundAtom],
+                            strips_ops: Sequence[STRIPSOperator],
+                            option_specs: Sequence[OptionSpec],
+                            ground_ops: Set[_GroundSTRIPSOperator]
+                            ) -> Callable[[Set[GroundAtom]], float]:
+        relaxed_operators = frozenset({utils.RelaxedOperator(
+            op.name, utils.atoms_to_tuples(op.preconditions),
+            utils.atoms_to_tuples(op.add_effects)) for op in ground_ops})
+        hadd_fn = utils.HAddHeuristic(
+            utils.atoms_to_tuples(init_atoms),
+            utils.atoms_to_tuples(goal),
+            relaxed_operators)
+        def _hadd_fn_h(init_atoms: Set[GroundAtom]) -> float:
```

Comment:
can you change `init_atoms` to something else here to avoid confusion with the thing in the outer scope?

Reply:
thanks missed that